
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Bioinformatics Workflow &#8212; Bioinformatics workflow for HT-SELEX analysis</title>
    
  <link href="_static/css/theme.css" rel="stylesheet">
  <link href="_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="_static/sphinx-book-theme.d59cb220de22ca1c485ebbdc042f0030.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Data visualisation" href="visualisation.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="_static/logo.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">Bioinformatics workflow for HT-SELEX analysis</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1 current">
  <a class="reference internal" href="#">
   Bioinformatics Workflow
  </a>
 </li>
</ul>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="visualisation.html">
   Data visualisation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="top_kmers.html">
   Top k-mers at SELEX cycle 6
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="coverage.html">
   Estimating optimal coverage for SELEX libraries
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="Notebook.html">
   Executable Notebook
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        <a class="dropdown-buttons"
            href="_sources/workflow.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download notebook file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="_sources/workflow.md"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.md</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
                onclick="printPdf(this)" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/kashyapchhatbar/eme_selex"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/kashyapchhatbar/eme_selex/issues/new?title=Issue%20on%20page%20%2Fworkflow.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Launch interactive content"><i class="fas fa-rocket"></i></button>
    <div class="dropdown-buttons">
        
        <a class="binder-button" href="https://mybinder.org/v2/gh/kashyapchhatbar/eme_selex/master?urlpath=tree/workflow.md"><button type="button"
                class="btn btn-secondary topbarbtn" title="Launch Binder" data-toggle="tooltip"
                data-placement="left"><img class="binder-button-logo"
                    src="_static/images/logo_binder.svg"
                    alt="Interact on binder">Binder</button></a>
        
        
        
        
    </div>
</div>

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show noprint">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#before-you-begin">
   Before you begin
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#preprocessing-of-fastq-files">
   Preprocessing of fastq files
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#quality-trim-sequencing-reads">
     Quality trim sequencing reads
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#automate-quality-trimming-using-a-workflow-manager">
     Automate quality trimming using a workflow manager
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#calculate-k-mer-abundance">
   Calculate k-mer abundance
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#saving-the-results">
     Saving the results
    </a>
   </li>
  </ul>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Bioinformatics Workflow</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#before-you-begin">
   Before you begin
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#preprocessing-of-fastq-files">
   Preprocessing of fastq files
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#quality-trim-sequencing-reads">
     Quality trim sequencing reads
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#automate-quality-trimming-using-a-workflow-manager">
     Automate quality trimming using a workflow manager
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#calculate-k-mer-abundance">
   Calculate k-mer abundance
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#saving-the-results">
     Saving the results
    </a>
   </li>
  </ul>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="bioinformatics-workflow">
<h1>Bioinformatics Workflow<a class="headerlink" href="#bioinformatics-workflow" title="Permalink to this headline">¶</a></h1>
<div class="section" id="before-you-begin">
<h2>Before you begin<a class="headerlink" href="#before-you-begin" title="Permalink to this headline">¶</a></h2>
<ul>
<li><p>Please install the following software in order to replicate the analysis.</p>
<ul class="simple">
<li><p><a class="reference external" href="https://github.com/seqan/flexbar"><code class="docutils literal notranslate"><span class="pre">flexbar</span></code></a> for quality trimming of sequencing reads</p></li>
<li><p><a class="reference external" href="https://snakemake.readthedocs.io/en/stable/"><code class="docutils literal notranslate"><span class="pre">snakemake</span></code></a> for workflow management</p></li>
<li><p><a class="reference external" href="https://github.com/kashyapchhatbar/every-motif-ever"><code class="docutils literal notranslate"><span class="pre">eme_selex</span></code></a> for kmer abundance</p></li>
<li><p><a class="reference external" href="https://jupyter.org/install"><code class="docutils literal notranslate"><span class="pre">jupyterlab</span></code></a> for web-based interactive development environment</p></li>
<li><p><a class="reference external" href="https://tqdm.github.io/"><code class="docutils literal notranslate"><span class="pre">tqdm</span></code></a> for displaying progress</p></li>
<li><p><a class="reference external" href="https://pandas.pydata.org/"><code class="docutils literal notranslate"><span class="pre">pandas</span></code></a> for data frame creation and manipulation</p></li>
<li><p><a class="reference external" href="https://seaborn.pydata.org/"><code class="docutils literal notranslate"><span class="pre">seaborn</span></code></a> for data visualisation</p></li>
<li><p><a class="reference external" href="https://plotly.com/python/getting-started/"><code class="docutils literal notranslate"><span class="pre">plotly</span></code></a> for data visualisation</p></li>
<li><p><a class="reference external" href="https://logomaker.readthedocs.io/"><code class="docutils literal notranslate"><span class="pre">logomaker</span></code></a> for making motif logos</p></li>
<li><p><a class="reference external" href="https://upsetplot.readthedocs.io/"><code class="docutils literal notranslate"><span class="pre">upsetplot</span></code></a> for visualising set overlaps</p></li>
</ul>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>Use <a class="reference external" href="https://anaconda.org/">anaconda</a> to create a separate environment and efficient installation. The following command will install all the relevant packages</p>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda create -n eme_selex -c bioconda flexbar snakemake pip
    jupyterlab tqdm pandas seaborn

conda activate eme_selex

pip install eme_selex logomaker upsetplot

conda install -c plotly <span class="nv">plotly</span><span class="o">==</span><span class="m">5</span>.6.0
</pre></div>
</div>
</li>
</ul>
</div>
<div class="section" id="preprocessing-of-fastq-files">
<h2>Preprocessing of fastq files<a class="headerlink" href="#preprocessing-of-fastq-files" title="Permalink to this headline">¶</a></h2>
<div class="section" id="quality-trim-sequencing-reads">
<h3>Quality trim sequencing reads<a class="headerlink" href="#quality-trim-sequencing-reads" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p>Remove low quality sequencing reads and trim them down to the number of bases in your random library. In our case, that number is 20. Use <a class="reference external" href="https://github.com/seqan/flexbar"><code class="docutils literal notranslate"><span class="pre">flexbar</span></code></a> with the following arguments below.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>flexbar --reads <span class="o">{</span>input<span class="o">}</span> --post-trim-length <span class="m">20</span> --min-read-length <span class="m">20</span> 
  --qtrim-threshold <span class="m">30</span> --output-reads <span class="o">{</span>output<span class="o">}</span> --fasta-output --number-tags 
  --stdout-log &gt; <span class="o">{</span>log<span class="o">}</span>
</pre></div>
</div>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>Use a workflow manager like <a class="reference external" href="https://snakemake.readthedocs.io/en/stable/"><code class="docutils literal notranslate"><span class="pre">snakemake</span></code></a> to automate this step for all samples.</p>
</div>
</li>
</ul>
</div>
<div class="section" id="automate-quality-trimming-using-a-workflow-manager">
<h3>Automate quality trimming using a workflow manager<a class="headerlink" href="#automate-quality-trimming-using-a-workflow-manager" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p>Using <a class="reference external" href="https://snakemake.readthedocs.io/en/stable/"><code class="docutils literal notranslate"><span class="pre">snakemake</span></code></a>, automate preprocessing of all samples. Create <code class="docutils literal notranslate"><span class="pre">Snakefile</span></code> in the parent directory, all the samples should be present in <code class="docutils literal notranslate"><span class="pre">fastq</span></code> sub-directory.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">samples</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;RV</span><span class="si">{</span><span class="n">s</span><span class="si">:</span><span class="s2">02d</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">40</span><span class="p">)]</span>

<span class="n">rule</span> <span class="n">preprocess</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span> <span class="n">expand</span><span class="p">(</span><span class="s2">&quot;fasta/</span><span class="si">{sample}</span><span class="s2">.fasta.gz&quot;</span><span class="p">,</span> <span class="n">sample</span><span class="o">=</span><span class="n">samples</span><span class="p">)</span>

<span class="n">rule</span> <span class="n">flexbar</span><span class="p">:</span>
    <span class="nb">input</span><span class="p">:</span> <span class="s2">&quot;fastq/</span><span class="si">{sample}</span><span class="s2">.fastq.gz&quot;</span>
    <span class="n">output</span><span class="p">:</span> <span class="s2">&quot;fasta/</span><span class="si">{sample}</span><span class="s2">.fasta.gz&quot;</span>
    <span class="n">log</span><span class="p">:</span> <span class="s2">&quot;logs/flexbar.</span><span class="si">{sample}</span><span class="s2">.log&quot;</span>
    <span class="n">shell</span><span class="p">:</span> <span class="s2">&quot;flexbar --reads </span><span class="si">{input}</span><span class="s2"> --post-trim-length 20 --min-read-length 20 </span>
              <span class="o">--</span><span class="n">qtrim</span><span class="o">-</span><span class="n">threshold</span> <span class="mi">30</span> <span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="n">reads</span> <span class="p">{</span><span class="n">output</span><span class="p">}</span> <span class="o">--</span><span class="n">fasta</span><span class="o">-</span><span class="n">output</span> <span class="o">--</span><span class="n">number</span><span class="o">-</span><span class="n">tags</span> 
              <span class="o">--</span><span class="n">stdout</span><span class="o">-</span><span class="n">log</span> <span class="o">&gt;</span> <span class="p">{</span><span class="n">log</span><span class="p">}</span><span class="s2">&quot;</span>
</pre></div>
</div>
</li>
<li><p>Execute the rule <code class="docutils literal notranslate"><span class="pre">flexbar</span></code> using the following command below.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>snakemake flexbar --cores <span class="m">1</span>
</pre></div>
</div>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>Increase the number of simultaneous executions by changing the parameter <code class="docutils literal notranslate"><span class="pre">--cores</span></code></p>
</div>
</li>
</ul>
</div>
</div>
<div class="section" id="calculate-k-mer-abundance">
<span id="kmer-fraction-from-file"></span><h2>Calculate k-mer abundance<a class="headerlink" href="#calculate-k-mer-abundance" title="Permalink to this headline">¶</a></h2>
<ul>
<li><p>Initialize <a class="reference external" href="https://docs.python.org/3/library/collections.html#collections.defaultdict"><code class="docutils literal notranslate"><span class="pre">defaultdict</span></code></a> to store the data.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">defaultdict</span>
<span class="n">counts</span><span class="p">,</span> <span class="n">fractions</span><span class="p">,</span> <span class="n">models</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">dict</span><span class="p">),</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">dict</span><span class="p">),</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">dict</span><span class="p">)</span>
</pre></div>
</div>
</li>
<li><p>Calculate kmer abundance, fraction and top position frequency matrix (PFM) models using <a class="reference external" href="https://github.com/kashyapchhatbar/every-motif-ever"><code class="docutils literal notranslate"><span class="pre">eme_selex</span></code></a>. For example, to calculate fraction of 5-mers, set <code class="docutils literal notranslate"><span class="pre">k=5</span></code> and execute the function over all samples.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">eme_selex.eme</span> <span class="kn">import</span> <span class="n">kmer_fraction_from_file</span> <span class="k">as</span> <span class="n">kf</span>
<span class="kn">from</span> <span class="nn">tqdm.auto</span> <span class="kn">import</span> <span class="n">tqdm</span>

<span class="n">samples</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;RV</span><span class="si">{</span><span class="n">s</span><span class="si">:</span><span class="s2">02d</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">40</span><span class="p">)]</span>

<span class="n">k</span> <span class="o">=</span> <span class="mi">5</span>

<span class="k">for</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;SampleName&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">leave</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="n">c</span><span class="p">,</span> <span class="n">f</span><span class="p">,</span> <span class="n">m</span> <span class="o">=</span> <span class="n">kf</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;fasta/</span><span class="si">{</span><span class="n">sample</span><span class="si">}</span><span class="s2">.fasta.gz&quot;</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="n">k</span><span class="p">)</span>
    <span class="n">counts</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span> <span class="o">=</span> <span class="n">c</span>
    <span class="n">fractions</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span> <span class="o">=</span> <span class="n">f</span>
    <span class="n">models</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span> <span class="o">=</span> <span class="n">m</span>
</pre></div>
</div>
</li>
</ul>
<div class="section" id="saving-the-results">
<h3>Saving the results<a class="headerlink" href="#saving-the-results" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p>Store the results in binary format for easy access and comparative analysis later.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pickle</span>
<span class="kn">import</span> <span class="nn">bz2</span>

<span class="k">with</span> <span class="n">bz2</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;results/results_</span><span class="si">{</span><span class="n">k</span><span class="si">}</span><span class="s2">.bz2&quot;</span><span class="p">,</span> <span class="s2">&quot;wb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">wH</span><span class="p">:</span>
    <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">counts</span><span class="p">,</span> <span class="n">wH</span><span class="p">)</span>
    <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">fractions</span><span class="p">,</span> <span class="n">wH</span><span class="p">)</span>
    <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">models</span><span class="p">,</span> <span class="n">wH</span><span class="p">)</span>
</pre></div>
</div>
</li>
</ul>
</div>
</div>
<div class="toctree-wrapper compound">
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            
                <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='right-next' id="next-link" href="visualisation.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Data visualisation</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            
        </div>
    </div>
    <footer class="footer">
  <p>
    
      By Kashyap Chhatbar<br/>
    
        &copy; Copyright 2021.<br/>
  </p>
</footer>
</main>


      </div>
    </div>
  
  <script src="_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>