{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3f558a7a",
   "metadata": {},
   "source": [
    "# Bioinformatics Workflow\n",
    "\n",
    "## Before you begin\n",
    "\n",
    "- Please install the following software in order to replicate the analysis.\n",
    "\n",
    "    - [`flexbar`](https://github.com/seqan/flexbar) for quality trimming of sequencing reads\n",
    "    - [`snakemake`](https://snakemake.readthedocs.io/en/stable/) for workflow management\n",
    "    - [`eme_selex`](https://github.com/kashyapchhatbar/every-motif-ever) for kmer abundance\n",
    "    - [`jupyterlab`](https://jupyter.org/install) for web-based interactive development environment \n",
    "    - [`tqdm`](https://tqdm.github.io/) for displaying progress\n",
    "    - [`pandas`](https://pandas.pydata.org/) for data frame creation and manipulation\n",
    "    - [`seaborn`](https://seaborn.pydata.org/) for data visualisation\n",
    "    - [`plotly`](https://plotly.com/python/getting-started/) for data visualisation\n",
    "    - [`logomaker`](https://logomaker.readthedocs.io/) for making motif logos\n",
    "    - [`upsetplot`](https://upsetplot.readthedocs.io/) for visualising set overlaps\n",
    "\n",
    "    ```{tip} \n",
    "    Use [anaconda](https://anaconda.org/) to create a separate environment and efficient installation. The following command will install all the relevant packages\n",
    "    ```\n",
    "\n",
    "    ```bash\n",
    "    conda create -n eme_selex -c bioconda flexbar snakemake pip\n",
    "        jupyterlab tqdm pandas seaborn\n",
    "\n",
    "    conda activate eme_selex\n",
    "\n",
    "    pip install eme_selex logomaker upsetplot\n",
    "\n",
    "    conda install -c plotly plotly==5.6.0\n",
    "    ```\n",
    "\n",
    "## Preprocessing of fastq files\n",
    "\n",
    "### Quality trim sequencing reads\n",
    "\n",
    "- Remove low quality sequencing reads and trim them down to the number of bases in your random library. In our case, that number is 20. Use [`flexbar`](https://github.com/seqan/flexbar) with the following arguments below.\n",
    "\n",
    "    ```bash\n",
    "    flexbar --reads {input} --post-trim-length 20 --min-read-length 20 \n",
    "      --qtrim-threshold 30 --output-reads {output} --fasta-output --number-tags \n",
    "      --stdout-log > {log}\n",
    "    ```\n",
    "\n",
    "    ```{tip} \n",
    "    Use a workflow manager like [`snakemake`](https://snakemake.readthedocs.io/en/stable/) to automate this step for all samples.\n",
    "    ```\n",
    "### Automate quality trimming using a workflow manager\n",
    "\n",
    "- Using [`snakemake`](https://snakemake.readthedocs.io/en/stable/), automate preprocessing of all samples. Create `Snakefile` in the parent directory, all the samples should be present in `fastq` sub-directory.\n",
    "\n",
    "    ```python\n",
    "    samples = [f\"RV{s:02d}\" for s in range(1, 40)]\n",
    "\n",
    "    rule preprocess:\n",
    "        input: expand(\"fasta/{sample}.fasta.gz\", sample=samples)\n",
    "\n",
    "    rule flexbar:\n",
    "        input: \"fastq/{sample}.fastq.gz\"\n",
    "        output: \"fasta/{sample}.fasta.gz\"\n",
    "        log: \"logs/flexbar.{sample}.log\"\n",
    "        shell: \"flexbar --reads {input} --post-trim-length 20 --min-read-length 20 \n",
    "                  --qtrim-threshold 30 --output-reads {output} --fasta-output --number-tags \n",
    "                  --stdout-log > {log}\"\n",
    "    ```\n",
    "\n",
    "- Execute the rule `flexbar` using the following command below.\n",
    "\n",
    "    ```bash\n",
    "    snakemake flexbar --cores 1\n",
    "    ```\n",
    "\n",
    "    ```{tip}\n",
    "    Increase the number of simultaneous executions by changing the parameter ``--cores``\n",
    "    ```\n",
    "(kmer_fraction_from_file)=\n",
    "## Calculate k-mer abundance\n",
    "\n",
    "- Initialize [`defaultdict`](https://docs.python.org/3/library/collections.html#collections.defaultdict) to store the data.\n",
    "\n",
    "    ```python\n",
    "    from collections import defaultdict\n",
    "    counts, fractions, models = defaultdict(dict), defaultdict(dict), defaultdict(dict)\n",
    "    ```\n",
    "\n",
    "- Calculate kmer abundance, fraction and top position frequency matrix (PFM) models using [`eme_selex`](https://github.com/kashyapchhatbar/every-motif-ever). For example, to calculate fraction of 5-mers, set `k=5` and execute the function over all samples.\n",
    "\n",
    "    ```python\n",
    "    from eme_selex.eme import kmer_fraction_from_file as kf\n",
    "    from tqdm.auto import tqdm\n",
    "\n",
    "    samples = [f\"RV{s:02d}\" for s in range(1, 40)]\n",
    "\n",
    "    k = 5\n",
    "\n",
    "    for sample in tqdm(df[\"SampleName\"].values, leave=False):\n",
    "        c, f, m = kf(f\"fasta/{sample}.fasta.gz\", k=k)\n",
    "        counts[sample] = c\n",
    "        fractions[sample] = f\n",
    "        models[sample] = m\n",
    "    ```\n",
    "\n",
    "### Saving the results\n",
    "\n",
    "- Store the results in binary format for easy access and comparative analysis later.\n",
    "\n",
    "    ```python\n",
    "    import pickle\n",
    "    import bz2\n",
    "\n",
    "    with bz2.open(f\"results/results_{k}.bz2\", \"wb\") as wH:\n",
    "        pickle.dump(counts, wH)\n",
    "        pickle.dump(fractions, wH)\n",
    "        pickle.dump(models, wH)\n",
    "    ```"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "formats": "md:myst",
   "text_representation": {
    "extension": ".md",
    "format_name": "myst"
   }
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  },
  "source_map": [
   11
  ]
 },
 "nbformat": 4,
 "nbformat_minor": 5
}